src/librustc_typeck/check/demand.rs

```diff
@@ -379,7 +379,23 @@ impl<'a, 'gcx, 'tcx> FnCtxt<'a, 'gcx, 'tcx> {
                     }
                 };
                 if self.can_coerce(ref_ty, expected) {
-                    if let Ok(src) = cm.span_to_snippet(sp) {
+                    let mut sugg_sp = sp;
+                    if let hir::ExprKind::MethodCall(segment, _sp, args) = &expr.node {
+                        let clone_trait = self.tcx.lang_items().clone_trait().unwrap();
+                        if let ([arg], Some(true), "clone") = (
+                            &args[..],
+                            self.tables.borrow().type_dependent_def_id(expr.hir_id).map(|did| {
+                                let ai = self.tcx.associated_item(did);
+                                ai.container == ty::TraitContainer(clone_trait)
+                            }),
+                            &segment.ident.as_str()[..],
+                        ) {
+                            // If this expression had a clone call when suggesting borrowing
+                            // we want to suggest removing it because it'd now be unecessary.
+                            sugg_sp = arg.span;
```

Comment:
You should also change `expr` (maybe better named `sugg_expr`, inside the `if let` below?) to `arg`, so that the `needs_parens` logic, and everything else, works correctly.
(e.g. I suspect you print `&1 + 2` instead of `&(1 + 2)` for `(1 + 2).clone()`)

Reply:
Nope, it works just fine:

```
error[E0308]: mismatched types
 --> src/main.rs:2:19
  |
2 |     let x: &i32 = (1i32 + 2).clone();
  |                   ^^^^^^^^^^^^^^^^^^
  |                   |
  |                   expected &i32, found i32
  |                   help: consider borrowing here: `&(1i32 + 2)`
  |
  = note: expected type `&i32`
             found type `i32`
```

Reply:
That's mildly disturbing - the code is *technically wrong*, since it's looking at the whole method call when determining whether it needs parens, but happens to produce the correct result.

Reply:
It's because the logic is to _add_ parentheses to the code when it is not already int he code. In this case the parentheses are already there so the expression doesn't need to change.

Reply:
`ExprKind::Paren` doesn't exist in the HIR, the first argument of the `clone` method call is `1i32 + 2` (which I would've assumed doesn't include parens in its own `Span`).